### PR TITLE
[CI] Exclude doc files from CI to save CI time

### LINKED ISSUE #3999: [CI] Don't run Formatting + main CI for `doc`
**Is your feature request related to a problem? Please describe.**

Currently `doc` is not built (in CI except tutorials). Any update in doc results in a unneeded compiles (6, 2 each for Windows, Mac and Ubuntu) wasting CI time

**Expected behavior**

Updates to doc, README, etc. only runs the relevant (or no) CI

**Current Behavior**

CI run at every push

**Describe alternatives you've considered**

Manually cancelling runs: doesn't communicate intent very well

**Good first issue** only if you know `azure`

### PR DESCRIPTION
**Pull request by ueqri**
Added the exclude rules for push and PR triggers. So updates to doc, README, etc. won't cause the formatting + main CI to run.

### CODE REVIEW

File: .ci/azure-pipelines/azure-pipelines.yaml
Comment (kunaltyagi) [inline .ci/azure-pipelines/azure-pipelines.yaml:7]: Does azure-pipelines support wildcard, like `*.md`?

Comment (ueqri) [reply .ci/azure-pipelines/azure-pipelines.yaml:7]: I referred this doc and wild cards are not supported with path filters. 
https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/azure-repos-git?view=azure-devops&tabs=yaml#paths

**Review by kunaltyagi — APPROVED** [review]

**Comment by kunaltyagi** [comment]
This PR still runs the main CI for docs. Create a new issue for the docs or edit the existing issue to reflect the partially resolved state?

**Comment by ueqri** [comment]
> This PR still runs the main CI for docs. Create a new issue for the docs or edit the existing issue to reflect the partially resolved state?

yeah, current CI build is still based on the old azure pipelines file in master. It will take effect after merging this PR.
Could I fix some small typo in pcl_style_guide I've found before to reflect the partially resolved state?

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
~Can't rescind my review from mobile, hence this. Please ignore~

Adding the rationale in a second review below

File: .ci/azure-pipelines/azure-pipelines.yaml
Comment (kunaltyagi) [inline .ci/azure-pipelines/azure-pipelines.yaml:4]: Adding this will result in the documentation stage not running.

The aim is to have the documentation CI run regardless, but the code CI (Ubuntu, Mac and Windows) to run only when code is changed (and doc is not code that's compiled in them)


I think properly implementing the feature will require one of the following
* adding a separate pipeline for documentation
* separate stage in other CI to use git to check which files are modified and skip if so

**Review by kunaltyagi — CHANGES_REQUESTED** [review]

File: .ci/azure-pipelines/azure-pipelines.yaml
Comment (ueqri) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: I'm sorry to cause some trouble, and increase your workload of review. I made the commit and assumed that any updates to doc(contains doc directory, README.md and etc) won't cause CI(includes documentation CI) to run at once. Whenever there is a commit with code files changed, the main CI(build and documentation) will run and the documentation repo will be updated by the way. 

Maybe that's not a nice way. And I think "adding a separate pipeline for documentation" will be a better choice and separating the main pipeline into pieces with clearer purposes will bring more convenience to CI. 

I'll do it soon. Thanks for your advice!

Comment (kunaltyagi) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: For separate CI, I don't know if it's possible to run a pipeline after another, but the following constraint would be nice to have:

The tutorials stage should be run after the documentation stage, only if the ubuntu build CI goes green.

Comment (ueqri) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: I follow the rules you commented.

In the latest commit, I separate both **documentation stage** and **tutorials stage** from **main build pipeline** to **documentation pipeline**. And I introduce [pipeline trigger](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/pipeline-triggers?view=azure-devops&tabs=yaml), which allows running a pipeline after another, to the new pipeline to ensure the tutorials stage will be triggered whenever the "Build GCC on Ubuntu" stage are completed successfully.

The expected behavior is listed below:
- code changes: **main build CI** run, and trigger the complete documentation CI only after "Build GCC" is OK.
- doc changes: only trigger documentation stage, and tutorials stage won't run. (which is current behavior of CI)

I've tested all those behaviors in [another branch](https://github.com/ueqri/pcl/commits/test-feature-on-minimal-CI) using simplified pipelines but the same structure.

And the result of the running is here: https://dev.azure.com/ueqri-ci/pcl-test/_build?view=runs

**Comment by ueqri** [comment]
I've implemented the new pipeline in `.ci/docs-pipeline.yaml`. If everything is ok, please add the pipeline in the Azure DevOps after merging. Thank you!

File: .ci/azure-pipelines/azure-pipelines.yaml
Comment (kunaltyagi) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: Thanks @ueqri 

Just a minor change:
* Tutorial stage should run in the end: regardless of Documentation pipeline or Build pipeline (since it serves 2 purposes: making sure tutorials work, with the latest code)

Apologies for not making it clear beforehand.

Situations:
1. Doc changes: format -> generate doc -> run tutorials stage
2. Code changes: format-> (run build, generate doc) -> run tutorials stage

For reference, currently, our pipeline performs (format -> build -> tutorials and doc in parallel) for every change.

I hope this clarifies the situation a bit.

Comment (ueqri) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: I think I got it, the tutorials stage should always run if everything works well(especially "build gcc on ubuntu").

BTW, I guess the "formatting stage" is unnecessary for docs pipeline, could I just run docs pipeline without it?

Comment (ueqri) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: I also find that the artifact(named "formatting") published by "formatting stage" has not been used in other files or pipelines. Is it an issue to be addressed or a feature as preparation for future work? Could I do something to help on this?

Comment (kunaltyagi) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: Ah. The thing is, we have not applied formatting to docs yet. The formatting is applied slowly across different modules (see the `.dev/format.sh` for details). Instead of modifying the pipeline later, we can have the required pipeline before we have formatting on docs.

>  Is it an issue to be addressed or a feature as preparation for future work? Could I do something to help on this?

It's just a patch that could be applied (useful, if we could somehow showcase it). I wanted a bot to auto-comment the diff, using the suggestion feature of GitHub, but it turned out that I don't have time for that 😆 That's something for a different PR though

Comment (ueqri) [reply .ci/azure-pipelines/azure-pipelines.yaml:4]: Thanks! I understand the structure of current CI better! 

I'll make the change for the following situations:
> 1. Doc changes: format -> generate doc -> run tutorials stage
> 2. Code changes: format-> (run build, generate doc) -> run tutorials stage

BTW, I'm interested in the bot to auto-comment the diff :laughing: and I want to learn and test it on my branch first.

**Comment by ueqri** [comment]
I think the latest commit looks good now. Could you take a look again please?

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (kunaltyagi) [inline .ci/azure-pipelines/docs-pipeline.yaml:7]: Why? :) No need to include these 3 md files

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (kunaltyagi) [inline .ci/azure-pipelines/docs-pipeline.yaml:35]: The build gcc step hasn't changed. Why will it trigger the docs pipeline?

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (ueqri) [reply .ci/azure-pipelines/docs-pipeline.yaml:7]: I apologize for that I've not explain my purpose well.

The expected behavior is:
- Docs change: only (push or PR) trigger docs pipeline(contains `formatting` now)
- Code change: only (push or PR) trigger main pipeline, and (pipeline) trigger docs pipeline, which means docs pipeline won't be triggered instantly only after `build_gcc` stage complete nicely.

It's a bit hard to implement flow control between different pipelines, so I use pipeline trigger to do this:
> The tutorials stage should be run after the documentation stage, only if the ubuntu build CI goes green.

To prevent the docs-pipeline running twice in code changes (the push&PR trigger, pipeline trigger), I restrict the push&PR trigger in docs-pipeline by adding `include` rules.

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (ueqri) [reply .ci/azure-pipelines/docs-pipeline.yaml:35]: Yes, it won't trigger `docs-pipeline` in docs changes. And I add the `condition` specially for code changes.

`docs-pipeline` triggers:
- by `build_gcc` stage only in code changes, so `formatting` has already run in `main-pipeline`, and it won't run here twice.
- by Push or PR only in docs changes(I include docs in `doc-pipeline`, and exclude them in `main-pipeline`), so `formatting` is ready to run in `docs-pipeline`.

**Comment by ueqri** [comment]
I admit that the method is kind of obscure and tricky now. I'd add git check to make it more legible and comprehensible.

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (kunaltyagi) [reply .ci/azure-pipelines/docs-pipeline.yaml:7]: Thanks for the explanation, but my point was that these 3 lines are not part of the documentation either, so no need to trigger the documentation pipeline

**Review by kunaltyagi — COMMENTED** [review]
Looks good.

@ueqri Could you do the following:
* Create a dummy project to show the CI runs (I believe you've already done this)
* Create a PR that runs only doc pipeline
* Create a PR that runs build then doc pipeline

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (ueqri) [reply .ci/azure-pipelines/docs-pipeline.yaml:7]: Ah...it's just my carelessness, thank you so much!!

**Comment by ueqri** [comment]
> Looks good.
> 
> @ueqri Could you do the following:
> 
>     * Create a dummy project to show the CI runs (I believe you've already done this)
> 
>     * Create a PR that runs only doc pipeline
> 
>     * Create a PR that runs build then doc pipeline

Of course:satisfied:, I will do it in my repository and paste the result here soon. Thank you:satisfied:!

**Comment by ueqri** [comment]
>     * Create a dummy project to show the CI runs (I believe you've already done this)
> 
>     * Create a PR that runs only doc pipeline
> 
>     * Create a PR that runs build then doc pipeline

I've made this in https://github.com/ueqri/pcl/tree/test-new-docs-pipeline, and related Azure pipelines dashboard are here: https://dev.azure.com/ueqri-ci/pcl-test/_build?view=runs

As I didn't set branch rules for triggers, some push beyond `test-new-docs-pipeline` branch also run CI and I canceled them immediately. So the result is a bit messy, I'm sorry for that.

The exactly related CI run are:

- PR that runs only doc pipeline: https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=108&view=results
- PR that runs build then doc pipeline: 
  - build pipeline: https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=113&view=results
  - triggered docs pipeline: https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=114&view=results
(name of triggered docs pipeline may be confused, but it does not matter, `Automatically triggered by Pipeline Build-CI/20210410.23` in result summary shows the pipeline works correctly)

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (kunaltyagi) [inline .ci/azure-pipelines/docs-pipeline.yaml:44]: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#succeeded

We can simplify using just `succeeded()` as the condition

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (kunaltyagi) [inline .ci/azure-pipelines/docs-pipeline.yaml:38]: `ne` here?
https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#ne

**Review by kunaltyagi — COMMENTED** [review]
The sample links look good.

Just a few minor comments

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (ueqri) [reply .ci/azure-pipelines/docs-pipeline.yaml:44]: Thanks for your reviews!

Actually, I've tried `succeeded()`,`always()` expressions in these commits to my test branch `test-new-docs-pipeline`: https://github.com/ueqri/pcl/commit/d4bb60aa4bfb0caf1c07ce82de08c6234598c2f2 https://github.com/ueqri/pcl/commit/691f2425a7f3ae6d6a7ced329e6d4e760285cc5b https://github.com/ueqri/pcl/commit/0ce26835e162b88711390f63e558966f7f944501

But the condition sometime works, and something failed. I guess those `succeeded()`,`always()` are actually `Job status check functions` https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#job-status-functions, but I'm not very sure and find no related official documents about applying job status check functions to stage check. So I turn to https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#dependencies, which was explicitly for stage check.

File: .ci/azure-pipelines/docs-pipeline.yaml
Comment (ueqri) [reply .ci/azure-pipelines/docs-pipeline.yaml:38]: Oh yes, L38 can be changed to `ne(in(dependencies.formatting.result, 'Failed', 'Canceled'))`, Thank you! :grinning:

Comment (kunaltyagi) [reply .ci/azure-pipelines/docs-pipeline.yaml:38]: Ah, my bad. Not there. I meant `ne` to replace the `not(eq(`, not here. 🤦🏾

Comment (kunaltyagi) [reply .ci/azure-pipelines/docs-pipeline.yaml:38]: Not here, but for L32

Comment (ueqri) [reply .ci/azure-pipelines/docs-pipeline.yaml:38]: I'll push the change immediately. Thanks for reviewing!

**Review by kunaltyagi — APPROVED** [review]

**Comment by ueqri** [comment]
I'm gratitude to you for helping me and guiding me in this PR. Thank you! :grinning: 

And I was wondering if this could be merged, and then I could contribute to this issue https://github.com/PointCloudLibrary/pcl/issues/3947 based on the changes in current branch. It's truly a great pleasure:heart:!

**Review by larshg — APPROVED** [review]

**Comment by kunaltyagi** [comment]
Note to maintainers: We'll need to update the list of required CI for merge

**Comment by ueqri** [comment]
I'm really sorry to bring this mistake in the last stage:cry:, and thanks for your hotfix. I promise I will check more carefully at any time especially in online code edits. I apologize for causing this trouble.

**Comment by kunaltyagi** [comment]
Meh. Happens :)

**Pull request merged**
